src/js/media-controller.js

```diff
@@ -38,357 +52,724 @@ const ButtonPressedKeys = ['ArrowLeft', 'ArrowRight', 'Enter', ' ', 'f', 'm', 'k
+    // For a more advanced solution, we could monitor for media state receivers that "care" about airplay support and add/remove
+    // whenever these are added/removed. (CJP)
+    // NOTE: I don't think only adding this if supported helps, so making this match others. 
+    // If Airplay's not supported then it it doesn't hurt to add it.
+    // If battery is really an issue we need a different approach.
+    // Also this is a terrible API, Apple. (heff)
+    get: function(controller, e){
+      if (!airplaySupported) return AvailabilityStates.UNSUPPORTED;
+
+      // NOTE: since we invoke all these event handlers without arguments whenever a media is attached,
+      // need to account for the possibility that event is undefined (CJP).
+      // TODO: Is there reall no way to detect this without an event? (heff)
+      if (!e) return undefined;
+      // TODO: Switch to non attr specific value
+
+      if (e.availability === 'available') {
+        return undefined;
+      } else if (e.availability === 'not-available') {
+        return AvailabilityStates.UNAVAILABLE;
+      }
+    },
+    mediaEvents: ['webkitplaybacktargetavailabilitychanged']
+  },
+  MEDIA_CAST_UNAVAILABLE: {
+    get: function() {
+      const castState = globalThis.CastableVideoElement?.castState;
+
+      if (!castSupported || !castState) {
+        return AvailabilityStates.UNSUPPORTED;
+      }
+
+      // Cast state: NO_DEVICES_AVAILABLE, NOT_CONNECTED, CONNECTING, CONNECTED
+      if (castState.includes('CONNECT')) {
+        // TODO: Move to non attr specific value
+        return undefined;
+      } else {
+        return AvailabilityStates.UNAVAILABLE;
+      }
+    },
+    mediaEvents: ['castchange']
+  },
+  MEDIA_FULLSCREEN_UNAVAILABLE: {
+    get: function() {
+      // TODO: Move to non attr specific value
+      return (fullscreenSupported) ? undefined : AvailabilityStates.UNAVAILABLE;
     }
-    if (!fullscreenEnabled) {
-      this._fullscreenUnavailable = AvailabilityStates.UNAVAILABLE;
+  },
+  MEDIA_PIP_UNAVAILABLE: {
+    get: function() {
+      // TODO: Move to non attr specific value
+      return (pipSupported) ? undefined : AvailabilityStates.UNSUPPORTED;
     }
-    if (!castSupported) {
-      this._castUnavailable = AvailabilityStates.UNSUPPORTED;
+  },
+  MEDIA_VOLUME_UNAVAILABLE: {
+    get: function(controller) {
+      if (volumeSupported !== undefined && !volumeSupported) {
+        return AvailabilityStates.UNSUPPORTED;
+      }
+
+      const { media } = controller;
+
+      if (media && typeof media.volume == 'undefined') {
+        return AvailabilityStates.UNAVAILABLE;
+      }
+
+      // TODO: move to non attr specific value
+      return undefined;
+    },
+    // Give a little time for the volume support promise to run
+    mediaEvents: ['loadstart']
+  },
+  MEDIA_CAPTIONS_LIST: {
+    get: function(controller) {
+      // TODO: Move to non attr specific values
```

Comment:
Should be fine to leave as is for now. These would get updated when we combine captions and subtitles lists.